### PR TITLE
Add GeoIP and GeoSite support for expr

### PR DESCRIPTION
**Pull request by KujouRinka**
I add geoip and geosite support:

```yaml
- name: geosite block bilibili
  action: block
  expr: geosite(string(tls?.req?.sni), "bilibili")

- name: geoip block some ip
  action: block
  expr: geoip(string(ip.dst), "cn")
```

But I think it's better to make this as a draft, for more things that might have further discussion:
- Most of files are directly copied from `github.com/apernet/hysteria/extras/acl`, I think only part of them are used so I decide to remove code that don't be used;
- I have no idea about could I organize file like this and file naming convention;
- More geo-like function could be added, such as CIDR support.

### CODE REVIEW

**Review by tobyxdd — COMMENTED** [review]
Much thanks for the PR, but I think there are a few issues:

- It seems like the entire ACL system from Hysteria is ported here? I'm not sure if we need all this except `matchers_v2geo.go` (and the `v2geo` package for parsing the data format). I think `geo_matcher.go` should contain the logic for creating and managing geoip/geosite matchers directly, instead of relying on `compileHostMatcher`, so we can remove the whole ACL compilation system, which contains a lot of code we don't need.
- Maybe we should put geoip/geosite & all future built-in functions into a subpackage of `ruleset`
- Need a way to expose config options for users to specify local geoip/geosite db files, similar to Hysteria.
- Correct me if I'm wrong but I think `GeoMatcher` has to be thread-safe? The ruleset can be called by multiple workers processing multiple streams at the same time.

**Comment by KujouRinka** [comment]
Here's what I do:
- Remove most of files but keep `v2geo` and `matchers_v2geo.go`;
- Rename `matcher.go` to `host_matchers.go` for I think this is helpful;
- I only keep `compileHostMatcher` function from `compile.go` and move it to `geo_matcher.go`;
- Add `Mutex` to protect field in `GeoMatcher`, but I think it may be better to use `RwLock` instead.

So it looks like:
```
ruleset
    ├── buildins
    │   └── geo
    │       ├── geo_loader.go
    │       ├── geo_matcher.go
    │       ├── host_matchers.go
    │       ├── interface.go
    │       ├── matchers_v2geo.go
    │       └── v2geo/
    ├── expr.go
    └── interface.go
```

If no further problem, I will continue working on this:
> Need a way to expose config options for users to specify local geoip/geosite db files, similar to Hysteria.

But may I ask that it's better to put this option to `config.yaml` in `rules.yaml`?

Thanks for your reading.

**Comment by tobyxdd** [comment]
👍 The changes look good but maybe remove `compileHostMatcher` entirely (extract the code that handles geoip/geosite into separate methods in `GeoMatcher`). btw "buildins" should be "builtins".

> But may I ask that it's better to put this option to config.yaml in rules.yaml?

In `config.yaml`, perhaps create a new section for ruleset settings as we may have more in the future

**Comment by tobyxdd** [comment]
```go
func NewGeoLoader() (*GeoMatcher, error) {
```

Is this a typo? It should be `NewGeoMatcher`

File: ruleset/buildins/geo/geo_matcher.go [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+
+type GeoMatcher struct {
+	geoLoader       GeoLoader
+	geoSiteMatcher  map[string]hostMatcher
+	siteMatcherLock sync.Mutex
+	geoIpMatcher    map[string]hostMatcher
+	ipMatcherLock   sync.Mutex
+}
+
+func NewGeoLoader() (*GeoMatcher, error) {
```
Comment (tobyxdd) [inline ruleset/buildins/geo/geo_matcher.go:18]: Is this a typo? I assume it should be `NewGeoMatcher`

File: ruleset/buildins/geo/host_matchers.go
Comment (tobyxdd) [inline ruleset/buildins/geo/host_matchers.go:1]: Do we need this file? I don't think we use any matchers from here

**Review by tobyxdd — CHANGES_REQUESTED** [review]

File: ruleset/buildins/geo/geo_matcher.go (hunk 1 above)
Comment (KujouRinka) [reply ruleset/buildins/geo/geo_matcher.go:18]: You are right.

File: ruleset/buildins/geo/host_matchers.go
Comment (KujouRinka) [reply ruleset/buildins/geo/host_matchers.go:1]: After removing some unnecessary code from `compileHostMatcher`, this file can be totally removed. I will rewrite and commit it later.

**Comment by KujouRinka** [comment]
To specify local geoip/geosite db files:

```yaml
# config.yaml

# other field ....

geo:
  geoip: geoip.dat
  geosite: geosite.dat
```

**Review by tobyxdd — CHANGES_REQUESTED** [review]

File: cmd/root.go [hunk 2]
```diff
@@ -161,6 +161,7 @@ func initLogger() {
 type cliConfig struct {
 	IO      cliConfigIO      `mapstructure:"io"`
 	Workers cliConfigWorkers `mapstructure:"workers"`
+	Geo     cliConfigRuleset `mapstructure:"ruleset"`
```
Comment (tobyxdd) [inline cmd/root.go:164]: Still named `Geo`

**Comment by KujouRinka** [comment]
I didn't find way to lazily build geosite and geoip by only calling `expr.Compile()` so I use `parser.Parse()` and `ast.Walk` to walk on AST before compile. I also move original `expr.Compile()` 's argument to `depVisitor` but I am not sure is this right.

https://github.com/KujouRinka/OpenGFW/blob/c26e04b5a6aa5ee7f4b9405702b873a0c4b4accd/ruleset/expr.go#L104-L123

I think geosite and geoip should differ from Analyzer so I add a function `func isBuiltInFunction(name string) bool`. If I am wrong please corrent me.

**Comment by tobyxdd** [comment]
Is it possible to always provide these functions when calling Compile, but only actually load the geo files when the identifier is seen by the visitor during compilation? I think this works the same, but can keep the code the way it was before

File: cmd/root.go [hunk 3]
```diff
@@ -161,7 +161,7 @@ func initLogger() {
 type cliConfig struct {
 	IO      cliConfigIO      `mapstructure:"io"`
 	Workers cliConfigWorkers `mapstructure:"workers"`
-	Geo     cliConfigRuleset `mapstructure:"ruleset"`
+	Geo     cliConfigGeo     `mapstructure:"ruleset"`
```
Comment (tobyxdd) [inline cmd/root.go:164]: Why did you rename it back... my previous comment was saying that you forgot to rename this variable to Ruleset, not asking you to change everything back to Geo 🤔

Comment (KujouRinka) [reply cmd/root.go:164]: Sorry for I misunderstood that.

**Comment by KujouRinka** [comment]
I agree with you, but I have a more question that loading geo files may counter with internet error, while `ast.Visitor` interface doesn't provide us how to return this. I think we should introduce a field in `depVisitor` for recording.

**Comment by tobyxdd** [comment]
Yes, feel free to modify depVisitor as you need.

**Comment by tobyxdd** [comment]
Or maybe just keep depVisitor a visitor (only record identifiers) and do the loading in the outer compile function?

**Comment by KujouRinka** [comment]
I rebase it from [1eb52f7](https://github.com/apernet/OpenGFW/pull/38/commits/1eb52f71af446376a09eab74152b4711341a3810) for making change clear.

**Comment by KujouRinka** [comment]
> Or maybe just keep depVisitor a visitor (only record identifiers) and do the loading in the outer compile function?

You mean mark whether geosite and geosite is needed in `depVisitor`, then after compile function to decide whether to load?

**Comment by KujouRinka** [comment]
I introduce `UseGeoSite` and `UseGeoIp` in `depVisitor` to mark this.

**Comment by KujouRinka** [comment]
I think I should rebase these commits from first commit of this PR right?

**Comment by tobyxdd** [comment]
> I think I should rebase these commits from first commit of this PR right?

What do you mean

**Comment by KujouRinka** [comment]
I think these commits are a bit messy, so before merging, maybe I need to use "git rebase" to merge all commits into one.

**Comment by tobyxdd** [comment]
That's fine, we can use squash merge.

**Comment by KujouRinka** [comment]
I think this draft is about to be completed so I convert it to a PR. If have any problem, please feel free to call me. ❤️

**Review by tobyxdd — APPROVED** [review]

**Comment by tobyxdd** [comment]
Thanks again for the effort. I plan to release v0.0.5 with this & WireGuard analyzer

**Pull request merged**